### DISCUSSION TITLE
How to enroll github provider from local development environment

### DISCUSSION THREAD

**Discussion by AGMETEOR** [stacklok/minder#1974] [General]
Error: `error_description: The redirect_uri MUST match the registered callback URL for this application.`
Hello. I am back again. So I have been trying to set up my local development environment and everything was going great until I hit a blocker.
I have created an OAuth App in my github account but I think the callback url referencing my local environment is wrong. The docs in the readme say it should point to `http://localhost:8081/realms/stacklok/broker/github/endpoint`. But when I try to enroll I get this page at `http://localhost:8081/realms/stacklok/broker/github/endpoint?error=redirect_uri_mismatch&error_description=The+redirect_uri+MUST+match+the+registered+callback+URL+for+this+application.&error_uri=https%3A%2F%2Fdocs.github.com%2Fapps%2Fmanaging-oauth-apps%2Ftroubleshooting-authorization-request-errors%2F%23redirect-uri-mismatch&state=AAAAAGWBxH5jhAH7anqA6mQH7KEpQtOVnqyuDqgGT7tS8WYhY64-cg`
<img width="1792" alt="Screenshot 2023-12-19 at 7 38 22 PM" src="https://github.com/stacklok/minder/assets/28563179/fb38464b-363c-496d-954d-c1084c025682">

Here's my app setup
<img width="1792" alt="Screenshot 2023-12-19 at 7 34 43 PM" src="https://github.com/stacklok/minder/assets/28563179/355bd648-af97-49c6-b070-db63e7fddcff">

**Comment by AGMETEOR** [comment]
After digging up. I found I had not added `http://localhost:8081/realms/stacklok/broker/github/endpoint` in `server-config.yaml`. It works now.

**Reply by lukehinds** [reply]
Awesome! Good to see you're up and running.

**Comment by rdimitrov** [comment]
Hey, @AGMETEOR 👋 

Glad to see it worked out 🚀

**Comment by AGMETEOR** [comment]
Hey @rdimitrov and @lukehinds thanks for looking at this discussion thread. I just have another issue on the enroll flow. After authorizing the app with extra permissions, the redirect to keycloak to this  url `http://localhost:8081/realms/stacklok/broker/github/endpoint/cli?code=b67c8df8825b3aa99c26&state=AAAAAGWC30-HMmo2iXesjcRU27IjyOykWBSCg8utcpbW-TUn5jTJmQ` seems to be broken locally. I get the page below

<img width="1792" alt="Screenshot 2023-12-20 at 3 34 39 PM" src="https://github.com/stacklok/minder/assets/28563179/21e0105d-c8ab-41e5-bead-feb5195efa49">

I am wondering if you have encountered this in your local development environment?

**Reply by rdimitrov** [reply]
Hey, so there are two OAuth apps we talk about in the docs:
1. One is for social login via GitHub - [Docs](https://minder-docs.stacklok.dev/run_minder_server/run_the_server#social-login) 
* The values from this one are used via `make github-login` to allow you to login via GitHub. 
3. The other one is for enrolling the provider - [Docs](https://minder-docs.stacklok.dev/run_minder_server/config_oauth#create-a-github-oauth-application)
* The values from this one are put in the `server-config.yaml` file

Note that the callback URL's are different for both, so you can check if you haven't mixed them up by any chance.

Disclaimer: Ignore the screenshots, we'll update them so they are up-to-date.

**Reply by AGMETEOR** [reply]
I think I definitely have things mixed up. The screenshots looking the same didn't help my case either :) I'll get back tomorrow with an update.

**Reply by AGMETEOR** [reply]
Ok finally managed to configure both OAuth apps and the enroll flow is now working well from local development. Thanks a lot for the clarification @rdimitrov

**Reply by rdimitrov** [reply]
Happy to hear that! 🙌
